### ISSUE TITLE
403 Error on repeat submission

### ISSUE THREAD

**Issue by Aryllen** [Sage-Bionetworks/dccvalidator#155]
There's an issue with the sessions in that some information must be being shared. When repeatedly submitting the same file, sometimes a 403 Error appears where it's thinking the file belongs to someone else.

![image](https://user-images.githubusercontent.com/25048342/67708390-51c90b80-f979-11e9-91ab-cb84d7389e60.png)

**Comment by karawoo** [comment]
Steps to reproduce as ascertained by running a [test app](https://github.com/karawoo/synapse-upload-test/blob/master/server.R) with @Aryllen and @kdaily:

1. Nicole visits app and uploads a file
2. Kenny visits app on a different machine
3. Nicole hits "submit" again -- error appears

**Comment by karawoo** [comment]
See the screenshot below, taken after step 3 described above:
<img width="1372" alt="Screen Shot 2019-10-29 at 2 02 26 PM" src="https://user-images.githubusercontent.com/4452678/67808907-df2d5e00-fa54-11e9-88b8-adc5953cd715.png">

Compare the error: 
```
403 Client Error: Only the creator of a FileHandle can access it directly by its ID. 
FileHandleId = '43863655', UserId = '3396533'
```

to the information about the file that was uploaded that's printed below:

```
createdBy[[1]] = 3368956 dataFileHandleId[[2]] = 43863655 modifiedBy[[3]] = 3368956
```

I am user 3368956, whoever logged in to the app in step 2 (Nicole in this case) is 3396533. The app knows that I was the one who created the file, but after Nicole has connected it won't let me update because it's recognizing her user ID instead of mine.

**Comment by brucehoff** [comment]
In `synapser` the user identity is 'global' to the package.  This allows you to have, for example:
```
synLogin(username, password)
synStore(somefile)
```
`synasper` then stores the file as the user given by 'username'.   If you use `synapser` for multiple users in the same R session then each call to a package function would have to be preceded by an invocation of `synLogin`, for example you should be calling `synLogin` here: https://github.com/karawoo/synapse-upload-test/blob/master/server.R#L63.  An alternative way to structure the package would be to create an object for each authenticated user and call the package functions through an instance of the object.   Our sense was that most analysts, working in a single-user environment, would not appreciate the more cumbersome syntax associated with that approach.  However the underlying *python* package does work that way and you could create multiple Python Synapse client objects through rPython, PythonInR or Reticulate.

**Comment by karawoo** [comment]
That makes sense, and is the same approach I took with [climbr](https://github.com/Sage-Bionetworks/climbr). While it would solve this problem, I could anticipate a lot of people being dissatisfied with such a major change, so I think we should discuss it with a wider group.

**Comment by karawoo** [comment]
After discussions today in the sysbio dcc meeting + with @ahayden separately, decided on the following:
1) As an immediate solution, we should try altering the shiny server configuration to create a new R process for each connection. We can use the [utilization scheduler](https://support.rstudio.com/hc/en-us/articles/220546267-Scaling-and-Performance-Tuning-Applications-in-Shiny-Server-Pro) and tweak [settings for individual apps](https://docs.rstudio.com/shiny-server/#how-do-i-translate-my-application-settings-to-nested-locations) within our shiny server configuration file, something like this:

```
location /users {

    # ...

    # Define rules to apply to kara's dccvalidator app
    location /kwoo/ShinyApps/dccvalidator-app {
        utilization_scheduler  1 0.7 10;
    }
}
```

I think this should work so long as the different R processes don't share a single python process. The same would need to be done for PsychENCODE's version of the app, and any others. It should be documented as part of #94.

2) In the longer term we'll move toward replacing synapser with the python client + reticulate, as this would also enable us to use python libraries in #72.

**Comment by Aryllen** [comment]
Since changing the shiny server configuration appears to be fixing this problem, could we close this issue, open a new issue for:

> 1. In the longer term we'll move toward replacing synapser with the python client + reticulate, as this would also enable us to use python libraries in #72.

and make a note in #94 for how to fix the problem for PsychENCODE?

**Comment by karawoo** [comment]
No, not yet. While our simple test app was fixed by making these changes to the shiny server, the real app was not. Xa is looking into it but we should keep this open for now.

**Comment by Aryllen** [comment]
Aww, okay. Too bad that it didn't fix the problem.

**Comment by apblair** [comment]
Hello,
I am running into the same error using the Synapse Python client version '1.9.4'. Please let me know what additional information I can provide.

**Comment by karawoo** [comment]
Hi @apblair, just so I'm understanding -- are you seeing this when using the Synapse python client directly, or when using the dccvalidator tool hosted in this GitHub repo? If you're seeing it in Python then I can try to pass on information to the appropriate people if you can tell me a little more about what you are trying to do when you encounter the error. Are you uploading the same files multiple times with different Synapse user accounts?
